src/contracts/core/AVSDirectory.sol

```diff
@@ -52,137 +61,662 @@ contract AVSDirectory is
+
+            emit OperatorAddedToOperatorSet(operator, OperatorSet({avs: avs, operatorSetId: operatorSetIds[i]}));
+        }
+    }
+
+    /**
+     * @notice Internal function to deregister an operator from an operator set.
+     *
+     * @param avs The AVS that the operator is deregistering from.
+     * @param operator The operator to deregister.
+     * @param operatorSetIds The IDs of the operator sets.
+     */
+    function _deregisterFromOperatorSets(address avs, address operator, uint32[] calldata operatorSetIds) internal {
+        // Loop over `operatorSetIds` array and deregister `operator` for each item.
+        for (uint256 i = 0; i < operatorSetIds.length; ++i) {
+            // Assert `operator` is registered for this iterations operator set.
+            require(
+                isMember[avs][operator][operatorSetIds[i]],
+                "AVSDirectory._deregisterOperatorFromOperatorSet: operator not registered for operator set"
+            );
+
+            // Mutate `isMember` to `false`.
+            isMember[avs][operator][operatorSetIds[i]] = false;
+
+            require(
+                operatorSetStatus[avs][operator][operatorSetIds[i]].registered,
+                "AVSDirectory._deregisterOperatorFromOperatorSet: operator not registered for operator set"
+            );
+            operatorSetStatus[avs][operator][operatorSetIds[i]] =
+                OperatorSetRegistrationStatus({registered: true, lastDeregisteredTimestamp: uint32(block.timestamp)});
+
+            emit OperatorRemovedFromOperatorSet(operator, OperatorSet({avs: avs, operatorSetId: operatorSetIds[i]}));
+        }
+    }
+
+    /**
+     * @notice For a single strategy, allocate magnitude to a list of operator sets and then update freeMagnitude
+     * for the strategy after all allocations
+     * @param operator address to allocate magnitude for
+     * @param allocation magnitude adjustment for a single strategy and multiple operator sets
+     * @param effectTimestamp timestamp when the allocation will take effect
+     */
+    function _allocate(address operator, MagnitudeAdjustment calldata allocation, uint32 effectTimestamp) internal {
+        IStrategy strategy = allocation.strategy;
+        OperatorSet[] calldata operatorSets = allocation.operatorSets;
+        uint64 freeAllocatableMagnitude = freeMagnitude[operator][strategy];
+
+        for (uint256 i = 0; i < operatorSets.length; ++i) {
+            // 1. check freeMagnitude available, that is the allocation of stake is backed and not slashable
+            require(
+                freeAllocatableMagnitude >= allocation.magnitudeDiffs[i],
+                "AVSDirectory.queueAllocations: insufficient available free magnitude to allocate"
+            );
+            // 2. rate limiting of 1 pending allocation at a time
+            (bool exists, uint32 timestamp, uint224 value) = _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i]
```

Comment:
we need to make this agnostic to the number of pending changes

Reply:
Something like below?

```
// read number of checkpoints after current timestamp
(uint224 value, uint256 pos) = _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i]
    .operatorSetId].upperLookupRecentWithPos(uint32(block.timestamp));
// value is 0 if no checkpoint exists at current timestamp
if (value != 0) {
    require(
        pos == _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId].length() - 1,
        "AVSDirectory.queueAllocations: only one pending allocation allowed for op, opSet, strategy"
    );
}
```